src/actions/mparticle/mparticle_transaction.ts

```diff
@@ -0,0 +1,287 @@
+      data.device_info = {}
+      data.custom_attributes = {}
+      if (mapping.eventName) {
+        Object.keys(mapping.eventName).forEach((attr: any) => {
+          data.event_name = row[attr].value
+        })
+      }
+      if (mapping.deviceInfo) {
+        Object.keys(mapping.deviceInfo).forEach((attr: any) => {
+          const key = mapping.deviceInfo[attr]
+          const val = row[attr].value
+          data.device_info[key] = val
+        })
+      }
+      if (mapping.dataEventAttributes) {
+        Object.keys(mapping.dataEventAttributes).forEach((attr: any) => {
+          const key = mapping.dataEventAttributes[attr]
+          const val = row[attr].value
+          data[key] = val
+        })
+      }
+      if (mapping.customAttributes) {
+        Object.keys(mapping.customAttributes).forEach((attr: any) => {
+          const key = mapping.customAttributes[attr]
+          const val = row[attr].value
+          data.custom_attributes[key] = val
+        })
+      }
+    }
+
+    return {
+      events: [
+        {
+          data,
+          event_type: EVENT_TYPE,
+        },
+      ],
+      user_attributes: eventUserAttributes,
+      user_identities: eventUserIdentities,
+      schema_version: 2,
+      environment: this.environment,
+    }
+  }
+
+  protected containsUserIdentity(userIdentities: any): boolean {
+    return (Object.getOwnPropertyNames(userIdentities).length > 0)
+  }
+
+  protected setEventType(dataType: string | undefined) {
+    if (dataType === USER) {
+      return USER
+    } else if (dataType === EVENT) {
+      return EVENT
+    }
+    throw "Missing data type (user|event)."
+  }
+
+  protected setEnvironment(env: string | undefined): string {
+    if (env === PROD_ENVIRONMENT) {
+      return PROD_ENVIRONMENT
+    }
+    return DEV_ENVIRONMENT
+  }
+
+  // Sets up the map object and loops over all fields.
+  protected createMappingFromFields(fields: any) {
+    let mapping: Mapping
+    if (this.eventType === USER) {
+      mapping = {
+        userIdentities: {},
+        userAttributes: {},
+        eventName: {},
+      }
+    } else {
+      mapping = {
+        userIdentities: {},
+        userAttributes: {},
+        eventName: {},
+        deviceInfo: {},
+        dataEventAttributes: {},
+        customAttributes: {},
+      }
+    }
+
+    fields.measures.forEach((field: ExploreField) => {
+      this.mapObject(mapping, field)
+    })
+    fields.dimensions.forEach((field: ExploreField) => {
+      this.mapObject(mapping, field)
+    })
+    fields.table_calculations.forEach((field: ExploreField) => {
+      this.mapObject(mapping, field)
+    })
+
+    if (!this.containsUserIdentity(mapping.userIdentities)) {
+      const err = "Each row must specify at least 1 identity tag."
+      this.errors.push(err)
+      throw err
+    }
+
+    return mapping
+  }
+
+  protected getTag(field: ExploreField): string {
+    // tslint:disable-next-line: strict-type-predicates
```

Comment:
The linter _really_ thinks these are always falsy / truthy (they're not), so I had to split them up and disable the strict type predicates here. I'm open to other suggestions to get it to work without disabling the linter, but I believe it's in error here

Reply:
```
  protected getTag(field: ExploreField): string {
    // tslint:disable-next-line: strict-type-predicates
    if (field.tags === undefined || field.tags[0] === undefined) {
      return ""
    }
    const tag = field.tags.find((t) => t.startsWith("mp_"))
    return tag ? tag : ""
  }
```
at least minimizes the lint disabling